### ISSUE TITLE
keyword_quote_linter() could get an allow_strings argument to allow "" quoting (not just backticks)

### ISSUE THREAD

**Issue by MichaelChirico** [r-lib/lintr#2053]
Follow-up to #2030.

Currently, the linter has the following behavior:

```r
1; c("a" = 1) # lints
2; c(`a` = 1) # lints
3; c("a b" = 1) # no lint
4; c(`a b` = 1) # no lint

5; "a" <- 1 # lint
6; `a` <- 1 # lint
7; "a b" <- 1 # lint
8; `a b` <- 1 # no lint

9; x$"a" # lint
10; x$`a` # lint
11; x$"a b" # lint
12; x$`a b` # no lint
```

This argument would allow customization on which of 3, 7, and 11 throw lints.

My thought for an API is

`keyword_quote_linter(allow_strings = c("formals", "assignment", "extraction"))` for cases 3, 7, and 11 respectively.

I would lean towards `allow_strings = character()` by default for consistency, and because `substitute(c("a b" = 1))` converts the quotes to backticks on printing:

```r
c("a b" = 1)
# c(`a b` = 1)
```

(for both assignment and extraction, `substitute()` leaves the delimiters alone)

**Comment by AshesITR** [comment]
I don't see why you'd want to want such fine-grained control and why you'd allow multiple variants of `` ` `` and `"`, or `'` at the same time. Maybe instead allow setting the canonical quote character, similar to `quotes_linter()`?
Default should be `` "`" `` of course.

**Comment by MichaelChirico** [comment]
For formals, there's an argument to allow quotes generally (even if the string needn't be quoted), e.g. since it used to be used in the dplyr documentation:

https://github.com/tidyverse/dplyr/blob/1cff4b9ce7f404955d8a5764b5708eee5972e0e3/R/join.r#L57

(the new `join_by()` syntax removed all those references)

We decided to allow `"` as an alternative to ` for formals only, mostly because it's just so common to see in documentation/elsewhere vs. backticks. So we are one user who benefits from the flexibility.
